### PR TITLE
Refactor/logic limits type

### PR DESCRIPTION
**Pull request by ccamel**
Implements #784 as discussed in [axone-protocol/discussions/2](https://github.com/orgs/axone-protocol/discussions/2).

- The properties `max_size`, `max_result_count`, `max_user_output_size`, and `max_variables` in the `Limits` message,  the properties `weighting_factor`, `default_predicate_cost` in the `GasPolicy` message, and `cost` in the `PredicateCost` message now use _uint64_ instead of `cosmossdk.io/math.Uint`. Similarly, the `limit` property in the `QueryServiceAskRequest` message has been updated to use _uint64_.
- Additionally, querying with a `limit` higher than `MaxResultCount` no longer results in an error. This ensures that clients still receive results while automatically adhering to the API’s limits, providing a smooth and efficient user experience.
- A migration of the `logic` module from v3 to v4 has been considered to ensure that the storage of these values is still functional in the [dentrite-1](https://github.com/axone-protocol/networks/blob/main/chains/dentrite-1/README.md) network.

An additional benefit is the reduced gas consumption, as the parameters being stored are smaller (see e2e tests).

## Tasks

- [x] update limits type to use _int64_.
- [x] update documentation
- [x] implement module migration
- [x] Migration has been tested on a _localnet_.

### localnet migration test

```sh
$ make chain-upgrade FROM_VERSION=v10.0.0 TO_VERSION=v11.0.0
...
2:21PM INF applying upgrade "v11.0.0" at height: 20 module=x/upgrade
2:21PM INF migrating module logic from version 3 to version 4 module=server
...
```

```sh
$ axoned query upgrade module-versions -o json | jq '.module_versions.[] | select(.name == "logic")'
{
  "name": "logic",
  "version": "4"
}
```

## Summary by CodeRabbit

- **New Features**
	- Streamlined handling of limit values in queries, enhancing performance.
	- Introduced new migration handler for upgrading store parameters.
	- Added support for gas policy configuration in parameters.
  
- **Bug Fixes**
	- Adjusted expected gas usage values across multiple scenarios to reflect optimizations.
  
- **Documentation**
	- Updated comments and documentation to clarify changes in parameter handling and migration processes.
  
- **Tests**
	- Expanded test coverage for bank and account predicates, including various edge cases.
	- Introduced new tests for migration logic to ensure correctness during version upgrades.

### CODE REVIEW

File: x/logic/util/pointer.go [hunk 1]
```diff
@@ -2,36 +2,18 @@ package util
...
-	return defaultValue
-}
-
-// IsNil returns true if the given value is nil, false otherwise.
-func IsNil(t any) bool {
-	return t == nil
+// IsZero returns true if the argument is nil or zero.
+func IsZero[T any](v T) bool {
+	v1 := reflect.ValueOf(v)
+	return !v1.IsValid() || v1.IsZero()
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/util/pointer.go:18]: This change is not part of the PR, as the code already existed. However, please open a ticket on GitHub to ensure your comment is considered for future improvements.

File: proto/logic/v1beta2/params.proto [hunk 2]
```diff
@@ -36,37 +36,21 @@ message Limits {
...
   // max_user_output_size specifies the maximum number of bytes to keep in the user output. If the user output exceeds
   // this size, the interpreter will overwrite the oldest bytes with the new ones to keep the size constant.
-  // nil value or 0 value means that no user output is used at all.
-  string max_user_output_size = 4 [
-    (gogoproto.moretags) = "yaml:\"max_user_output_size\"",
-    (gogoproto.customtype) = "cosmossdk.io/math.Uint",
-    (gogoproto.nullable) = true
-  ];
+  // A value of 0 means the user output is disabled.
+  uint64 max_user_output_size = 4 [(gogoproto.moretags) = "yaml:\"max_user_output_size\""];
```
[the remark this answers is not in the document]
Comment (ccamel) [reply proto/logic/v1beta2/params.proto:49]: I believe it’s quite clear, as the comment above already explains:

> // max_user_output_size specifies the maximum number of bytes to keep in the user output. If the user output exceeds
> // this size, the interpreter will overwrite the oldest bytes with the new ones to keep the size constant.

Therefore, if this size is set to 0, it indicates that no output will be retained.

File: x/logic/keeper/interpreter.go [hunk 3]
```diff
@@ -210,6 +210,17 @@ func lookupCost(predicate string, defaultCost uint64, costs []types.PredicateCos
...
+// calculateSolutionLimit returns the final number of solutions to be returned based on the given number of solutions and the
+// maximum result count.
+func calculateSolutionLimit(nbSolutions uint64, maxResultCount uint64) uint64 {
+	nbSolutions = max(nbSolutions, 1)
+	if maxResultCount == 0 {
+		return nbSolutions
+	}
+
+	return min(nbSolutions, maxResultCount)
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/keeper/interpreter.go:222]: ah, that’s not correct. _min_ and _max_ **are** built-in functions in Go 1.21 (see [here](https://olontsev.io/new-min-and-max-built-in-functions-in-go-1-21/)).

**Comment by ccamel** [comment]
@dependabot pause

File: x/logic/util/prolog.go [hunk 4]
```diff
@@ -34,14 +33,14 @@
 	}
 
 	var env *engine.Env
-	count := sdkmath.ZeroUint()
-	envs := make([]*engine.Env, 0, sdkmath.MinUint(solutionsLimit, sdkmath.NewUint(defaultEnvCap)).Uint64())
+	count := uint64(0)
+	envs := make([]*engine.Env, 0, min(solutionsLimit, defaultEnvCap))
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/util/prolog.go:37]: That’s not correct. min and max are built-in functions in Go 1.21 (see [here](https://olontsev.io/new-min-and-max-built-in-functions-in-go-1-21/)).

File: x/logic/types/params.go [hunk 5]
```diff
@@ -127,30 +127,30 @@
...
+		i.MaxUserOutputSize = maxUserOutputSize
 	}
 }
 
 // WithMaxVariables sets the maximum number of variables that can be created by the interpreter.
-func WithMaxVariables(maxVariables math.Uint) LimitsOption {
+func WithMaxVariables(maxVariables uint64) LimitsOption {
 	return func(i *Limits) {
-		i.MaxVariables = &maxVariables
+		i.MaxVariables = maxVariables
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/types/params.go:153]: Nope. The full range of `uint64`, including `zero`, is accepted (a value of 0 means that there is no limit on the size of the program).

File: x/logic/types/params.go [hunk 6]
```diff
@@ -185,3 +173,37 @@
...
+
+// NewGasPolicy creates a new GasPolicy object.
+func NewGasPolicy(opts ...GasPolicyOption) GasPolicy {
+	g := GasPolicy{}
+	for _, opt := range opts {
+		opt(&g)
+	}
+
+	return g
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/types/params.go:209]: Honestly, I don’t think it’s very valuable. This code is pur technical, and already covered by other functional tests, and the documentation doesn’t provide any additional insight beyond what’s already available in the API documentation.

File: Makefile [hunk 7]
```diff
@@ -249,49 +249,39 @@ chain-upgrade: build ## Test the chain upgrade from the given FROM_VERSION to th
...
 	\
 	sleep 5;\
  	$$BINARY_OLD tx gov vote 1 yes \
      		--from validator \
      		--yes \
      		--home ${CHAIN_HOME} \
      		--chain-id axone-${CHAIN} \
-     		--keyring-backend test \
-     		-b sync; \
+     		--keyring-backend test; \
```
[the remark this answers is not in the document]
Comment (ccamel) [reply Makefile:284]: Yes, you’re right, but in this case, I’m just updating the existing code to align with dependency updates. I don’t want to make any further changes in this PR since the makefile goal’s purpose is simply to facilitate local migration tests. It’s used in very specific cases, infrequently, and by people who already know what they’re doing.

File: app/upgrades.go [hunk 8]
```diff
@@ -1,17 +1,28 @@
...
+}
+
+// registerUpgradeHandlers registers the chain upgrade handlers.
+func (app *App) registerUpgradeHandlers() {
+	for _, upgrade := range upgrades {
+		app.UpgradeKeeper.SetUpgradeHandler(
+			upgrade,
+			func(ctx context.Context, _ upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
+				return app.ModuleManager.RunMigrations(ctx, app.configurator, vm)
+			},
```
[the remark this answers is not in the document]
Comment (ccamel) [reply app/upgrades.go:23]: AFAIS, it's already managed by the sdk. So there's no need to do anything more.

File: x/logic/keeper/migrations.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+		}
+		if v := paramsFrom.Limits.MaxResultCount; v != nil {
+			paramsTo.Limits.MaxResultCount = v.Uint64()
+		}
+		if v := paramsFrom.Limits.MaxUserOutputSize; v != nil {
+			paramsTo.Limits.MaxUserOutputSize = v.Uint64()
+		}
+		if v := paramsFrom.Limits.MaxVariables; v != nil {
+			paramsTo.Limits.MaxVariables = v.Uint64()
+		}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/keeper/migrations.go:50]: Nope. `paramsFrom.Limts` is not a pointer.

File: x/logic/keeper/migrations.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+			paramsTo.GasPolicy.DefaultPredicateCost = v.Uint64()
+		}
+		if v := paramsFrom.GasPolicy.PredicateCosts; v != nil {
+			for _, pc := range v {
+				paramsTo.GasPolicy.PredicateCosts = append(paramsTo.GasPolicy.PredicateCosts, types.PredicateCost{
+					Predicate: pc.Predicate,
+					Cost:      pc.Cost.Uint64(),
+				})
+			}
+		}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply x/logic/keeper/migrations.go:66]: Nope. `paramsFrom.GasPolicy` is not a pointer.

File: Makefile [hunk 11]
```diff
@@ -249,49 +249,39 @@
...
+	    --summary "⬆️ Upgrade the chain from ${FROM_VERSION} to ${TO_VERSION}" \
+	    --upgrade-height 20 \
+	    --upgrade-info "{}" \
+	    --deposit 10000000uaxone \
+	    --no-validate \
+	    --yes \
+	    --from validator \
+	    --keyring-backend test \
+	    --chain-id axone-${CHAIN} \
+	    --home ${CHAIN_HOME}; \
```
[the remark this answers is not in the document]
Comment (ccamel) [reply Makefile:275]: Well, it’s all good—it’s just a Makefile goal to facilitate migration test, and there’s nothing critical about it. In my opinion, the values have been set reasonably.

File: app/app.go [hunk 12]
```diff
@@ -814,9 +814,9 @@
 		panic(err)
 	}
 
-	// RegisterUpgradeHandlers is used for registering any on-chain upgrades.
+	// registerUpgradeHandlers is used for registering any on-chain upgrades.
 	// Make sure it's called after `app.ModuleManager` and `app.configurator` are set.
-	app.RegisterUpgradeHandlers()
+	app.registerUpgradeHandlers()
```
[the remark this answers is not in the document]
Comment (ccamel) [reply app/app.go:819]: Don’t worry, everything is under control. We have test coverage automatically verified by the CI. 😌

File: app/upgrades.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ccamel) [reply app/upgrades.go:23]: Don’t worry, everything is under control. We have test coverage automatically verified by the CI. 😌

**Comment by ccamel** [comment]
@coderabbitai pause

**Review by amimart — APPROVED** [review]
Awesome! And thanks a lot for repairing the `chain-upgrade` Markefile target 😉
